pkg/log/zap/flags.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package zap contains helpers for setting up a new logr.Logger instance
+// using the Zap logging framework.
+
+package zap
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/spf13/pflag"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+type encoderConfigFunc func(*zapcore.EncoderConfig)
+
+type encoderFlag struct {
+	setFunc func(zapcore.Encoder)
+	value   string
+}
+
+var _ pflag.Value = &encoderFlag{}
+
+func (ev *encoderFlag) String() string {
+	return ev.value
+}
+
+func (ev *encoderFlag) Type() string {
+	return "encoder"
+}
+
+func (ev *encoderFlag) Set(flagValue string) error {
+	val := strings.ToLower(flagValue)
+	switch val {
+	case "json":
+		ev.setFunc(newJSONEncoder())
+	case "console":
+		ev.setFunc(newConsoleEncoder())
+	default:
+		return fmt.Errorf("invalid encoder value \"%s\"", flagValue)
+	}
+	ev.value = flagValue
+	return nil
+}
+
+func newJSONEncoder(ecfs ...encoderConfigFunc) zapcore.Encoder {
```

Comment:
```suggestion
func newJSONEncoder() zapcore.Encoder {
```
As @joelanford suggested let's remove the `encoderConfigFunc` type for now.

I think we'll first want to make the [EncoderConfig](https://godoc.org/go.uber.org/zap/zapcore#EncoderConfig) more generally configurable via fields in the Options struct, and helper functions to set those options. 
And in a way so they don't get overridden by other Option fields like Development.

Then based on the fields in the Options struct that set EncoderConfig fields (e.g timestamp format and maybe more) we can add flags to configure those.

If it's simple enough to do, we could add it in this PR but I think a follow up might be better if not.